### PR TITLE
provide++

### PR DESCRIPTION
**Pull request by bennn**
Alternate provide form. Requires a contract and documentation. Only works for identifiers.

Sample:

```
(provide/api
  square
  #:contract (-> number? number?)
  #:doc "`(square n)` computes `n^2`"
  pad
  #:contract (->* [string? natural?] [#:fill-char char?] string?)
  #:doc "`(pad str n #:fill-char c)` creates a string of length `n` whose first
    characters match `str` and remaining characters are duplicates of `c`.
    By default, `#:fill-char` is `#\space`"
)
```

To discuss:
- [x] do we like the choice of keywords? Want to change, or add options?
- [x] do we like the order of keywords, or should that be flexible
- [x] okay to include all of `racket/contract` when requiring `mechanics`?
- [x] need to `provide/api` more than just identifiers? (We can always still `provide` in a file.)

Lower priority (create issues for these)
- [ ] pick a syntax for disabling contacts
- [ ] decide how / where to compile documentation
- [ ] create a "docstring" language (or syntax class)

### CODE REVIEW

**Comment by bkc39** [comment]
> do we like the choice of keywords? Want to change, or add options?

Keywords look good to me. Maybe consider `#:required-contracts 'all` as the default or you could provide a list with the identifiers that require a docstring. Likewise try `#:required-docstrings 'all` as the default.

> do we like the order of keywords, or should that be flexible

I suppose it's easy enough to change them around, so why not? I think it's fine the way they are though.

> okay to include all of racket/contract when requiring mechanics

I never really use anything outside of `racket/contract/base`. I think reduce the import to that. But I require it in all my files any way to `provide` the contracts.

> need to provide/api more than just identifiers? (We can always still provide in a file)

I have never `provided` anything that wasn't an identifier. I suppose this is a wishlist feature. 

> pick a syntax for disabling contacts

I like the idea of

``` racket
(provide/api #:no-contracts '(some-identifier another-identifier)
  ;; This guy needs the contract
  third-indentifier
  #:doc "docstring"
  #:contract (-> any any)
  ;; doc still required here
  some-identifier
  #:doc "still required"
  ...)
```

By default all docstrings and contracts are required. I also think it might be better to group them with brackets

``` racket
(provide/api #:no-contracts '(some-identifier another-identifier)
  ;; This guy needs the contract
  [third-indentifier
   #:doc "docstring"
   #:contract (-> any any)]
  ;; doc still required here
  [some-identifier
   #:doc "still required"]
  ...)
```

> decide how / where to compile documentation

My opinion on this issue is that the docstring is for the toplevel, and we use Scribble for official documentation. So something like

``` racket
;; something.rkt
(provide/api
  something
  #:doc "A very important thing"
  #:contract (-> any any))
```

Then in the REPL:

``` racket
> (require "something.rkt")
> (something-doc)
A very important thing
```

It would also be nice to be able to generate `.scrbl` files automatically from the docstrings.

> create a "docstring" language (or syntax class)

I would be totally cool with having this be just inline Scribble.

Basically, I really like the `numpy` style.

``` python
In [1]: numpy.sin?
String form:     <ufunc 'sin'>
File:            /tp64/python/2.7.2/lib/python2.7/site-packages/numpy/__init__.py
Docstring:
sin(x[, out])

Trigonometric sine, element-wise.

Parameters
----------
x : array_like
    Angle, in radians (:math:`2 \pi` rad equals 360 degrees).

Returns
-------
y : array_like
    The sine of each element of x.

See Also
--------
arcsin, sinh, cos

Notes
-----
The sine is one of the fundamental functions of trigonometry
(the mathematical study of triangles).  Consider a circle of radius
1 centered on the origin.  A ray comes in from the :math:`+x` axis,
makes an angle at the origin (measured counter-clockwise from that
axis), and departs from the origin.  The :math:`y` coordinate of
the outgoing ray's intersection with the unit circle is the sine
of that angle.  It ranges from -1 for :math:`x=3\pi / 2` to
+1 for :math:`\pi / 2.`  The function has zeroes where the angle is
a multiple of :math:`\pi`.  Sines of angles between :math:`\pi` and
:math:`2\pi` are negative.  The numerous properties of the sine and
related functions are included in any standard trigonometry text.

Examples
--------
Print sine of one angle:

>>> np.sin(np.pi/2.)
1.0

Print sines of an array of angles given in degrees:

>>> np.sin(np.array((0., 30., 45., 60., 90.)) * np.pi / 180. )
array([ 0.        ,  0.5       ,  0.70710678,  0.8660254 ,  1.        ])

Plot the sine function:

>>> import matplotlib.pylab as plt
>>> x = np.linspace(-np.pi, np.pi, 201)
>>> plt.plot(x, np.sin(x))
>>> plt.xlabel('Angle [rad]')
>>> plt.ylabel('sin(x)')
>>> plt.axis('tight')
>>> plt.show()
```

I'd love to have that same functionality for out library (probably with much less thorough documentation). I think that test cases should automatically be turned into examples. A stripped down example

``` racket
(provide/api
  fact
  #:contract (-> exact-nonnegative-integer? exact-nonnegative-integer?)
  #:doc
  "
  Returns the factorial of a number.

  Parameters
  ----------

  n : A positive integer.

  Returns
  -------

  The factorial of n

  Examples
  ---------

  > (fact 5)
  120
  ")

(define (fact n) (if (= n 0) 1 (* n (fact (sub1 n)))))
```

Or alternatively, using s-expressions:

``` racket
(provide/api
  fact
  #:contract (-> exact-nonnegative-integer? exact-nonnegative-integer?)
  #:doc
  (docstring
    (summary "Returns the factorial of a number")
    (parameters
     [n "A positive Integer."])
    (returns "The factorial of `n`")
    (examples
      ('(fact 5) . 120))))

(define (fact n) (if (= n 0) 1 (* n (fact (sub1 n)))))
```

The latter definitely feels for "Racket-y", but I'd be fine with either.

**Comment by bennn** [comment]
1. Using `contract/base` is a good point, I forgot there were variants of normal `racket/contract`. 
2. I want `#:contract` and `#:doc` on every function -- they're always required. What I meant was disabling the contract check. `#:nocontracts` is a good idea. My idea was replacing `#:contract` with `#:dummy-contract` or something. But `provide/api/nocheck` might be simplest.

Hmm. We really want `no-check` on the user-side, not the library side. I guess we should ignore this for now.
1. `id-doc` looks great, and that should be easy to do. We might also be able to attach the doc to the actual object or fill a `help()` hashtable, but I'm not sure how to implement those.
2. The numpy string is too big for me. I'd like `provide/api` to be the racket answer to `.mli` files. The S-expression is nice though, and that leaves it abstract for things like `summary` to expand to ascii or latex or scribble.

Immediate plans:
- [x] change to `contract/base`
- [ ] export an `id-doc` string or function for all provides

**Comment by bkc39** [comment]
I was just thinking that we can probably implement "turning off" contracts in the same way as [`#lang typed/racket/nocheck`](http://docs.racket-lang.org/ts-reference/Typed_Racket_Syntax_Without_Type_Checking.html).

**Comment by bennn** [comment]
Hmm, maybe we can just patch racket with a `#lang racket/no-contracts`. We should look into how no-check is implemented already.

**Comment by bennn** [comment]
Just changed the macro to accept an optional boolean with the `#:contract` keyword

```
(provide/api 
  ;; Provided with a contract
  a #:contract (-> integer? integer?) #:doc "something"
  b #:contract #t (-> integer? integer?) #:doc "another thing"
  ;; No contract
  c #:contract #f (-> integer? integer?) #:doc "an unsafe thing"
)
...
```

**Comment by bennn** [comment]
I'm still not sure how to use `#:doc`. I don't want the macro to define and export an identifier. I don't think syntax properties will appear in the requiring file.
